### PR TITLE
Finishing up reporting feature

### PR DESCRIPTION
**Pull request by julianguyen**
# Description 

Finishing up reporting feature which @aSquare14 and @prateksha worked on for Rails Girls Summer of Code 2018 🎉 

### What I've done

- [x] Add missing test coverage
- [x] Send banned user email notifying whether they've been banned or unbanned
- [x] Cleanup UI
- [x] Fix linting
- [x] Fix translation bugs
- [x] Cleanup ReportsController
- [x] Re-add support for Comments
- [x] Ran `annotate` since there were schema changes

### In another PR
- Notify admin user when new report is submitted

## Additional Details

Do `Rebase and Merge` once pull request is ready to merge

## Corresponding Issue

https://github.com/ifmeorg/ifme/issues/14

# Screenshots

capture

image

image

image

# Test Coverage

✅

### CODE REVIEW

File: config/locales/es.yml [hunk 1]
```diff
@@ -336,6 +353,14 @@ es:
...
       join_meeting_subject: 'if me | %{user} se ha unido a "%{meeting}" de "%{group}"'
       join_meeting_body: '<p>Para ver "%{meeting}", %{link}!</p>'
+      reported_user_subject: 'if me | Usted ha informado de un usuario'
+      reported_user_body: 'Usted ha informado %{reported_name}. Esto será revisado por un administrador.'
+      reportee_user_subject: 'if me | Has sido reportado'
+      reportee_user_body: 'Has sido reportado. Esto será revisado por un administrador.'
+      add_ban_subject: 'if me | Has sido prohibido'
+      add_ban_body: 'Has sido prohibido por violar a nuestra %{code_of_conduct_link}.'
+      remove_ban_subject: 'if me | Ya no estas prohibido'
+      remove_ban_body: 'Usted fue previamente prohibido por violar a nuestra %{code_of_conduct_link}. Ya no estás prohibido.'
```
Comment (alvarocasadoc) [inline config/locales/es.yml]: I would rephrase this to something like "Anteriormente fuiste baneado/a por violar nuestro %{code_of_conduct_link}. Ya no estás baneado/a" or "El ban aplicado sobre tu cuenta por violar nuestro %{code_of_conduct_link} ha sido cancelado".

File: app/controllers/profile_controller.rb [hunk 2]
```diff
@@ -11,4 +11,45 @@ def index
...
+    return unless user.any?
+
+    if banned
+      BannedMailer.add_ban_email(user.first).deliver_now
+    else
+      BannedMailer.remove_ban_email(user.first).deliver_now
+    end
+  end
+
+  def ban_or_remove(banned)
```
Comment (nshki) [inline app/controllers/profile_controller.rb]: I wonder if there's a more expressive name we can give this method. The two wrappers above (`ban_user` and `remove_ban`) make this very clear, but as a standalone method, I had to do a double take. Calling `ban_or_remove(false)` is a little confusing, but that could also just be me.

Would something like `ban` be too generic?

File: app/controllers/profile_controller.rb [hunk 3]
```diff
@@ -11,4 +11,45 @@ def index
     @stories = Kaminari.paginate_array(get_stories(user))
                        .page(params[:page])
   end
+
+  def ban_user
+    ban_or_remove(true)
+  end
+
+  def remove_ban
```
Comment (nshki) [inline app/controllers/profile_controller.rb]: Would like to propose `unban_user` as the action name here. Or alternatively, renaming `ban_user` to `add_ban`.

File: app/views/profile/index.html.erb [hunk 4]
```diff
@@ -36,6 +39,17 @@
         </div>
       <% end %>
     <% end %>
+    <% if current_user.admin? %>
+      <% if @profile.banned?%>
```
Comment (nshki) [inline app/views/profile/index.html.erb]: Missing space before `%>`.

**Comment by julianguyen** [comment]
Woo, thank you for the reviews @alvarocasadoc and @nshki 🎉 Especially for improved Spanish translations!

File: app/controllers/profile_controller.rb [hunk 5]
```diff
@@ -11,4 +11,45 @@ def index
...
+    ban(false)
+  end
+
+  private
+
+  def ban_mailer(user, banned)
+    return unless user.any?
```
Comment (arun1595) [inline app/controllers/profile_controller.rb]: The argument `user`, contrary to what it says, actually holds an `ActiveRecord_Relation` object which is a bit confusing imo. What do you think of calling the `ban_mailer` method with the `User` object instead? This way, `ban_mailer` doesn't need to check if it was passed a non-empty collection and it doesn't need to get the user object with the `user.first` call.

The `ban` and `ban_mailer` methods would look something like

```ruby
def ban(banned)
  return unless current_user.admin
  user = User.find_by(id: params[:user_id])
  return unless user

  user.update(banned: banned)
  ban_mailer(user, banned)
  redirect_to(
    admin_dashboard_path,
    notice_or_alert(
      user,
      banned ? 'user_banned' : 'ban_removed'
    )
  )
end

def ban_mailer(user, banned)
  if banned
    BannedMailer.add_ban_email(user).deliver_now
  else
    BannedMailer.remove_ban_email(user).deliver_now
  end
end
```

And we could go even further and move all the ban and remove ban logic into a Service object.

Thoughts?

File: app/controllers/profile_controller.rb [hunk 6]
```diff
@@ -11,4 +11,45 @@ def index
...
+
+    if banned
+      BannedMailer.add_ban_email(user.first).deliver_now
```
Comment (arun1595) [inline app/controllers/profile_controller.rb]: Do we really need to pass the whole user object to the mailer method?

AFAIK, the arguments passed to the mailer method will be serialized and stored in the queue to be processed later. Serializing a heavy `User` object is expensive and we should avoid it . Someone let me know if I'm wrong.

I suggest we pass the `user_id` to the mailer method and let it do the job of looking up the `User` object.

@julianguyen  What do you think?

File: app/controllers/profile_controller.rb (hunk 5 above)
Comment (julianguyen) [reply app/controllers/profile_controller.rb]: This is a really good call out! Thank you. I think for now we can keep it in the controller, but I'll update things so that the user object is actually the expected type.

File: app/controllers/profile_controller.rb (hunk 6 above)
Comment (julianguyen) [reply app/controllers/profile_controller.rb]: That's a good suggestion, will update :D Thanks!

File: app/helpers/notification_mailer_helper.rb [hunk 7]
```diff
@@ -20,7 +20,7 @@ def comment_on_strategy_subject(data)
 
   def comment_on_meeting_subject(data)
     group_id = Meeting.find(data['typeid']).group_id
-    group = Group.where(id: group_id).first.name
+    group = Group.find_by(id: group_id).name
```
Comment (hmasila) [inline app/helpers/notification_mailer_helper.rb]: 👍

File: app/helpers/notification_mailer_helper.rb [hunk 8]
```diff
@@ -51,10 +51,10 @@ def new_group_body(data)
       val('new_group_body'),
       user: data['user'].to_s,
       group: data['group'].to_s,
-      description: Group.where(id: data['group_id']).first.description,
+      description: Group.find_by(id: data['group_id']).description,
```
Comment (hmasila) [inline app/helpers/notification_mailer_helper.rb]: 👍 I hope there will always be a group with this id `data['group_id']` otherwise we will have to deal with `no method error description for nil class`

Comment (julianguyen) [reply app/helpers/notification_mailer_helper.rb]: There should be, but I can double check and fix it in another PR (since it's technically out of scope). Good call out, thanks :)
